File: app/src/main/java/pl/laskowski/marcin/type/ClassicSquare.java
package pl.laskowski.marcin.type;

import pl.laskowski.marcin.model.Sudoku;

/**
 * Created by Marcin Laskowski.
 */

public class ClassicSquare extends ClassicRectangle {

    public ClassicSquare(int size) {
        super(size, size, blockSize(size), blockSize(size));
    }

    private static int blockSize(int size) {
        int blockSize = (int) Math.sqrt(size);
        if (blockSize * blockSize != size) throw new IllegalArgumentException();
        return blockSize;
    }

    @Override
    public Sudoku template() {
        return Sudoku.blank(width(), height());
    }

}

File: app/src/main/java/pl/laskowski/marcin/type/ClassicSquare.kt
package pl.laskowski.marcin.type

import kotlin.math.sqrt

class ClassicSquare(size: Int) : ClassicRectangle(size, size, blockSize(size), blockSize(size)) {

    companion object {
        private fun blockSize(size: Int): Int {
            val blockSize = sqrt(size.toDouble()).toInt()
            require(blockSize * blockSize == size)
            return blockSize
        }
    }
}